In [1]:
import pandas as pd
import numpy as np

In [2]:
#loading the dataset
anime_data = pd.read_csv(r'C:\Users\JADESOLA\Documents\AI\MadeWithML\Recommendation system\datasets\Anime_cleaned.csv')

In [3]:
anime_data = anime_data.rename(columns={'Anime_id': 'anime_id', 'Title': 'title','Genre':'genres','Synopsis':'synopsis',
                                       'Type':'anime_type','Producer':'producer','Studio':'studio','Rating':'rating','ScoredBy':'scored_by','Popularity':'popularity',
                                       'Members':'members','Episodes':'num_of_episodes','Source':'anime_source'})

In [4]:
#creating the function to remove the characters
import re
def clean_up(s):
  new_s = re.sub(r"[^\sa-zA-Z0-9\.\,\;]","",str(s))
  return new_s

In [5]:
anime_data['title'] = anime_data['title'].apply(clean_up)
anime_data['genres'] = anime_data['genres'].apply(clean_up)
anime_data['title'] = anime_data['title'].apply(clean_up)
anime_data['producer'] = anime_data['producer'].apply(clean_up)
anime_data['studio'] = anime_data['studio'].apply(clean_up)
anime_data['synopsis'] = anime_data['synopsis'].apply(clean_up)

In [6]:
anime_data.head(3)

anime_id                           title  \
0         1                    Cowboy Bebop   
1         5  Cowboy Bebop Tengoku no Tobira   
2         6                          Trigun   

                                        genres  \
0    Action,Adventure,Comedy,Drama,SciFi,Space   
1             Action,Space,Drama,Mystery,SciFi   
2  Action,SciFi,Adventure,Comedy,Drama,Shounen   

                                            synopsis anime_type  \
0  In the year 2071, humanity has colonized sever...         TV   
1  Another day, another bountysuch is the life of...      Movie   
2  Vash the Stampede is the man with a 60,000,000...         TV   

               producer    studio  rating  scored_by  popularity   members  \
0          BandaiVisual   Sunrise    8.81   363889.0        39.0  704490.0   
1  Sunrise,BandaiVisual     Bones    8.41   111187.0       475.0  179899.0   
2   VictorEntertainment  Madhouse    8.31   197451.0       158.0  372709.0   

   num_of_episodes anime_source                        aired  \
0             26.0     Original  Apr 3, 1998 to Apr 24, 1999   
1              1.0     Original                  Sep 1, 2001   
2             26.0        Manga  Apr 1, 1998 to Sep 30, 1998   

                                                Link  
0       https://myanimelist.net/anime/1/Cowboy_Bebop  
1  https://myanimelist.net/anime/5/Cowboy_Bebop__...  
2             https://myanimelist.net/anime/6/Trigun

In [7]:
anime_data.shape

(15832, 15)

In [8]:
anime_data.dtypes

anime_id             int64
title               object
genres              object
synopsis            object
anime_type          object
producer            object
studio              object
rating             float64
scored_by          float64
popularity         float64
members            float64
num_of_episodes    float64
anime_source        object
aired               object
Link                object
dtype: object

In [9]:
anime_data.isnull().sum()

anime_id              0
title                 0
genres                0
synopsis              0
anime_type          634
producer              0
studio                0
rating             2501
scored_by          3700
popularity          619
members               0
num_of_episodes    2825
anime_source       1927
aired               619
Link                619
dtype: int64

In [10]:
#dropping the null rows
anime_data.dropna(axis = 0,how = 'any',inplace=True)

In [11]:
anime_data.shape

(11658, 15)

# Content based using genre, producer, source and type

In [12]:
# Apply clean_data function to your features.
features = ['genres', 'producer', 'studio', 'anime_source','anime_type']

for feature in features:
    anime_data[feature] = anime_data[feature].apply(clean_up)

In [13]:

anime_data.head()

anime_id                           title  \
0         1                    Cowboy Bebop   
1         5  Cowboy Bebop Tengoku no Tobira   
2         6                          Trigun   
3         7              Witch Hunter Robin   
4         8                  Bouken Ou Beet   

                                           genres  \
0       Action,Adventure,Comedy,Drama,SciFi,Space   
1                Action,Space,Drama,Mystery,SciFi   
2     Action,SciFi,Adventure,Comedy,Drama,Shounen   
3  Action,Magic,Police,Supernatural,Drama,Mystery   
4          Adventure,Fantasy,Shounen,Supernatural   

                                            synopsis anime_type  \
0  In the year 2071, humanity has colonized sever...         TV   
1  Another day, another bountysuch is the life of...      Movie   
2  Vash the Stampede is the man with a 60,000,000...         TV   
3  Witches are individuals with special powers li...         TV   
4  It is the dark century and the people are suff...         TV   

               producer          studio  rating  scored_by  popularity  \
0          BandaiVisual         Sunrise    8.81   363889.0        39.0   
1  Sunrise,BandaiVisual           Bones    8.41   111187.0       475.0   
2   VictorEntertainment        Madhouse    8.31   197451.0       158.0   
3          BandaiVisual         Sunrise    7.34    31875.0      1278.0   
4                   nan  Toei Animation    7.04     4757.0      3968.0   

    members  num_of_episodes anime_source                         aired  \
0  704490.0             26.0     Original   Apr 3, 1998 to Apr 24, 1999   
1  179899.0              1.0     Original                   Sep 1, 2001   
2  372709.0             26.0        Manga   Apr 1, 1998 to Sep 30, 1998   
3   74889.0             26.0     Original   Jul 2, 2002 to Dec 24, 2002   
4   11247.0             52.0        Manga  Sep 30, 2004 to Sep 29, 2005   

                                                Link  
0       https://myanimelist.net/anime/1/Cowboy_Bebop  
1  https://myanimelist.net/anime/5/Cowboy_Bebop__...  
2             https://myanimelist.net/anime/6/Trigun  
3  https://myanimelist.net/anime/7/Witch_Hunter_R...  
4     https://myanimelist.net/anime/8/Bouken_Ou_Beet

In [14]:
#creating a function to split the particular columns into lists again
def split_columns(x):
  return x.split(',')

In [15]:
#applying the function to the same set of columns
for feature in features:
    anime_data[feature] = anime_data[feature].apply(split_columns)

In [16]:
anime_data.head(3)

anime_id                           title  \
0         1                    Cowboy Bebop   
1         5  Cowboy Bebop Tengoku no Tobira   
2         6                          Trigun   

                                              genres  \
0   [Action, Adventure, Comedy, Drama, SciFi, Space]   
1             [Action, Space, Drama, Mystery, SciFi]   
2  [Action, SciFi, Adventure, Comedy, Drama, Shou...   

                                            synopsis anime_type  \
0  In the year 2071, humanity has colonized sever...       [TV]   
1  Another day, another bountysuch is the life of...    [Movie]   
2  Vash the Stampede is the man with a 60,000,000...       [TV]   

                  producer      studio  rating  scored_by  popularity  \
0           [BandaiVisual]   [Sunrise]    8.81   363889.0        39.0   
1  [Sunrise, BandaiVisual]     [Bones]    8.41   111187.0       475.0   
2    [VictorEntertainment]  [Madhouse]    8.31   197451.0       158.0   

    members  num_of_episodes anime_source                        aired  \
0  704490.0             26.0   [Original]  Apr 3, 1998 to Apr 24, 1999   
1  179899.0              1.0   [Original]                  Sep 1, 2001   
2  372709.0             26.0      [Manga]  Apr 1, 1998 to Sep 30, 1998   

                                                Link  
0       https://myanimelist.net/anime/1/Cowboy_Bebop  
1  https://myanimelist.net/anime/5/Cowboy_Bebop__...  
2             https://myanimelist.net/anime/6/Trigun

In [17]:
# Function to convert all strings to lower case and strip names of spaces
def clean_data(x):
    if isinstance(x, list):
        return [str.lower(i.replace(" ", "")) for i in x]
    else:
        #Check if source exists. If not, return empty string
        if isinstance(x, str):
            return str.lower(x.replace(" ", ""))
        else:
            return ''

In [18]:
#applying the clean_data function
for feature in features:
    anime_data[feature] = anime_data[feature].apply(clean_data)

In [19]:
def create_soup(x):
    return ' '.join(x['genres']) + ' ' +' '.join(x['producer']) + ' ' + ' '.join(x['studio']) + ' '+\
                ' '.join(x['anime_source']) + ' '  + ' '.join(x['anime_type'])

In [20]:
# Create a new soup feature
anime_data['soup'] = anime_data.apply(create_soup, axis=1)

In [21]:
anime_data['soup'].head()

0    action adventure comedy drama scifi space band...
1    action space drama mystery scifi sunrise banda...
2    action scifi adventure comedy drama shounen vi...
3    action magic police supernatural drama mystery...
4    adventure fantasy shounen supernatural nan toe...
Name: soup, dtype: object

In [22]:
# Import CountVectorizer and create the count matrix
from sklearn.feature_extraction.text import CountVectorizer

count = CountVectorizer(stop_words='english')
count_matrix = count.fit_transform(anime_data['soup'])

We make use of the CountVectorizer library. This is to ensure there is no down weighting of any of the elemnets. Even if appeared more than once it will be considered. This will give us a matrix where each column represents a word in the overview vocabulary (all the words that appear in at least one document), and each row represents an anime, as before.

In [23]:
count_matrix.shape

(11658, 1251)

From the above, there are 1251 vocabularies in the metadata that you fed to it.
Next, we will use the cosine_similarity to measure the distance between the embeddings.

We create a function to get top 10 similar animes, based on it's features

In [24]:
# Compute the Cosine Similarity matrix based on the count_matrix
from sklearn.metrics.pairwise import cosine_similarity

cosine_sim1 = cosine_similarity(count_matrix, count_matrix)

We create a function to get top 10 similar animes, based on it's features

In [25]:
#Construct a reverse map of indices and anime titles
indices = pd.Series(anime_data.index, index=anime_data['title']).drop_duplicates()

In [26]:
indices.head()

title
Cowboy Bebop                      0
Cowboy Bebop Tengoku no Tobira    1
Trigun                            2
Witch Hunter Robin                3
Bouken Ou Beet                    4
dtype: int64

In [85]:
# Function that takes in anime title as input and outputs most similar anime
def get_recommendations(title, cosine_sim=cosine_sim1):
    # Get the index of the anime that matches the title
    idx = indices[title]
    
    # Get the pairwsie similarity scores of all anime with that anime
    sim_scores = list(enumerate(cosine_sim[idx]))
    
    # Sort the anime based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    
    # Get the scores of the 25 most similar anime
    sim_scores = sim_scores[1:26]
    
    #converting the sim_scores to a dictionary
    dict_scores = dict(sim_scores)
    
    # Get the anime indices
    #anime_indices = [i for i in dict_scores]
    #print(anime_indices)

    #ensuring recommendations don't have the same name with title input
    final_indices = []
    for i in dict_scores:
        if title not in anime_data['title'].iloc[i]:
            final_indices.append(i)
        else:
            pass
   
    # Return the top 10 most similar anime
    final_indices = final_indices[0:11]
    return anime_data[['title','genres']].iloc[final_indices]

In [87]:
recommendations = get_recommendations('Naruto')
print(recommendations)

                                   title  \
239                               Bleach   
212                       Rekka no Honoo   
9171                    D.Grayman Hallow   
10599     Boruto Jump Festa 2016 Special   
3453                     Dragon Ball Kai   
6799                Dragon Ball Kai 2014   
10300                       Black Clover   
1333   Bleach Movie 1 Memories of Nobody   
2423                  Asobot Senki Gokuu   
5572              Ninkuu Knife no Bohyou   
11449   Boku no Hero Academia 3rd Season   

                                                  genres  
239    [action, adventure, comedy, superpower, supern...  
212    [action, adventure, martialarts, shounen, supe...  
9171    [action, adventure, superpower, demons, shounen]  
10599  [action, adventure, comedy, superpower, martia...  
3453   [action, adventure, comedy, fantasy, martialar...  
6799   [action, adventure, comedy, fantasy, martialar...  
10300          [action, comedy, magic, fantasy, shounen]  

# Recommendation based on storyline

In [44]:
anime_data['synopsis'].head()

0    In the year 2071, humanity has colonized sever...
1    Another day, another bountysuch is the life of...
2    Vash the Stampede is the man with a 60,000,000...
3    Witches are individuals with special powers li...
4    It is the dark century and the people are suff...
Name: synopsis, dtype: object

Applying vectorization to the plot summary to get other similar animes, using tf-idf vectorizer

In [46]:
#converting the synopsis column to word vector
from sklearn.feature_extraction.text import TfidfVectorizer

# Using Abhishek Thakur's arguments for TF-IDF
tfv = TfidfVectorizer(min_df=3,  max_features=None, 
            strip_accents='unicode', analyzer='word',token_pattern=r'\w{1,}',
            ngram_range=(1, 3), use_idf=1,smooth_idf=1,sublinear_tf=1,
            stop_words = 'english')

# Filling NaNs with empty string
anime_data['synopsis'] = anime_data['synopsis'].fillna('')

# Fitting the TF-IDF on the 'overview' text
tfv_matrix = tfv.fit_transform(anime_data['synopsis'])

tfv_matrix.shape

(11658, 26977)

With this matrix in hand, we compute similarity scores using cosine similarity

In [49]:
#calculating similarity scores
from sklearn.metrics.pairwise import linear_kernel

# Compute the cosine similarity matrix using linear kernel
cos_sim2 = linear_kernel(tfv_matrix, tfv_matrix)

We define a function that takes in an anime title as an input and outputs a list of the 10 most similar animes. Firstly, for this, you need a reverse mapping of anime titles and DataFrame 

In [50]:
# Reverse mapping of indices and movie titles
indices = pd.Series(anime_data.index, index=anime_data['title']).drop_duplicates()

In [66]:
get_recommendations('Naruto', cos_sim2) #based on storyline

title  \
11320              Sylvanian Families Mini Story   
1003   Saint Seiya Saishuu Seisen no Senshitachi   
149                                Mononoke Hime   
157                                      Tactics   
5648     Higurashi no Naku Koro ni Kaku Outbreak   
9198                                     Mayoiga   
6659    Toukiden The Age of Demons  Introduction   
7770                              Bakemono no Ko   
5529                 Kamisama no Inai Nichiyoubi   
7925                                    Paniponi   
11918                          Yoninme no Ousama   

                                                  genres  
11320                                [kids, sliceoflife]  
1003                         [adventure, scifi, shounen]  
149                         [action, adventure, fantasy]  
157    [mystery, comedy, historical, demons, supernat...  
5648          [horror, mystery, psychological, thriller]  
9198     [mystery, comedy, horror, psychological, drama]  
6659                                    [action, demons]  
7770                           [adventure, supernatural]  
5529                                  [fantasy, mystery]  
7925                                 [sliceoflife, kids]  
11918                                      [music, kids]

In [69]:
#writing a function to get recommendations using both storyline and features based on 50% for both
def get_recommendations_both(title, cos_sim1, cos_sim2):
    # Get the index of the movie that matches the title
    idx = indices[title]

    # Get the pairwsie similarity scores of all anime using similarity1
    sim_scores1 = list(enumerate(cos_sim1[idx]))

    # Get the pairwsie similarity scores of all anime using similarity1
    sim_scores2 = list(enumerate(cos_sim2[idx]))

    #Getting the average of both similarity scores
    sim_scores_avg = [(sim_scores1[i][0],(sim_scores1[i][1] + sim_scores2[i][1])/2) for i in range(len(sim_scores1))]

    # Sort the movies based on the similarity scores
    sim_scores_avg = sorted(sim_scores_avg, key=lambda x: x[1], reverse=True)

    #Get the scores of the 10 most similar movies
    sim_scores_avg = sim_scores_avg[1:26]

    #Get the movie indices
    anime_indices = [i[0] for i in sim_scores_avg]
      
    #ensuring recommendations don't have the same name with title input
    final_indices = []
    for i in anime_indices:
        if title not in anime_data['title'].iloc[i]:
            final_indices.append(i)
        else:
            pass

    # Return the top 10 most similar anime
    final_indices = final_indices[0:11]
    return anime_data[['title','genres']].iloc[final_indices]


In [70]:
get_recommendations_both('Naruto',cosine_sim1,cos_sim2)

title  \
239                              Bleach   
9171                   D.Grayman Hallow   
212                      Rekka no Honoo   
10599    Boruto Jump Festa 2016 Special   
10300                      Black Clover   
3453                    Dragon Ball Kai   
5572             Ninkuu Knife no Bohyou   
6799               Dragon Ball Kai 2014   
11449  Boku no Hero Academia 3rd Season   
8480                  Dragon Ball Super   
2423                 Asobot Senki Gokuu   

                                                  genres  
239    [action, adventure, comedy, superpower, supern...  
9171    [action, adventure, superpower, demons, shounen]  
212    [action, adventure, martialarts, shounen, supe...  
10599  [action, adventure, comedy, superpower, martia...  
10300          [action, comedy, magic, fantasy, shounen]  
3453   [action, adventure, comedy, fantasy, martialar...  
5572   [action, adventure, comedy, martialarts, shounen]  
6799   [action, adventure, comedy, fantasy, martialar...  
11449      [action, comedy, school, shounen, superpower]  
8480   [action, adventure, comedy, fantasy, martialar...  
2423         [action, scifi, adventure, comedy, shounen]